contracts/PooledStaking.sol

```diff
@@ -0,0 +1,215 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/
+*/
+
+pragma solidity ^0.5.16;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/upgrades/contracts/Initializable.sol";
+import "./base/MasterAware.sol";
+
+contract PooledStaking is MasterAware {
+  using SafeMath for uint;
+
+  enum ParamType {
+    MIN_DEPOSIT_AMOUNT,
+    MIN_STAKE_PERCENTAGE,
+    MAX_LEVERAGE,
+    UNSTAKE_LOCK_TIME
+  }
+
+  // Minimum deposit
+  uint public MIN_DEPOSIT_AMOUNT;
+
+  // Minimum allowed stake percentage per contract
+  uint public MIN_STAKE_PERCENTAGE;
+
+  // Sum of all stakes should not exceed the total deposited amount times this number
+  uint public MAX_LEVERAGE;
+
+  // Lock period before unstaking takes place
+  uint public UNSTAKE_LOCK_TIME;
+
+  struct Staker {
+    uint staked; // total amount of staked nxm
+    uint reward; // total amount that is ready to be claimed
+    address[] contracts; // list of contracts the staker has staked on
+
+    // Percentage of staked NXM amount on a contract
+    // All percentages have 2 decimal places, i.e. 100% is represented as 10000, and 33.33% as 3333
+    mapping(address => uint) allocations;
+
+    // amount to be subtracted after all deallocations will be processed
+    mapping(address => uint) deallocationAmounts;
+
+    // pending deallocations
+    mapping(address => uint[]) pendingDeallocationIds;
+  }
+
+  struct Contract {
+    uint staked; // amount of nxm staked for this contract
+    address[] stakers; // used for iteration
+  }
+
+  struct DeallocationRequest {
+    address contractAddress;
+    address stakerAddress;
+    uint amount;
+    uint deallocateAt;
+  }
+
+  // List of all contract addresses
+  address[] public contractAddresses;
+
+  // List of all staker addresses
+  // address[] public stakerAddress;
+
+  // stakers mapping. stakerAddress => Staker
+  mapping(address => Staker) public stakers;
+
+  // contracts mapping. contractAddress => Staker
+  mapping(address => Contract) public contracts;
+
+  // deallocation requests mapping. contractAddress => Staker
+  mapping(address => DeallocationRequest) public deallocationRequests;
+
+  uint firstDeallocation;
+  uint lastDeallocation;
+
+  function initialize(address masterAddress) initializer public {
+    MasterAware.initialize(masterAddress);
+  }
+
+  function stake(uint amount) onlyMembers external {
+
+    require(amount > MIN_DEPOSIT_AMOUNT, "Amount is less than minimum allowed");
+
+    Staker storage staker = stakers[msg.sender];
+    uint oldStake = staker.staked;
+    staker.staked = staker.staked.add(amount);
+
+    for (uint i = 0; i < staker.contracts.length; i++) {
```

Comment:
i think we should document with a comment why this loop is bounded, based on the parameters MIN_STAKE_PERCENTAGE and MAX_LEVERAGE as discussed.

Reply:
ok this is part of a view now